### PR TITLE
Make normal self-modifying code work

### PR DESCRIPTION
**Pull request by rhelmot**
#266 was patched incorrectly. The fix made it so that the entire block would be re-lifted instead of just a single instruction, which discards the relevant self-modifications. The real fix is to keep an explicit flag for whether we've actually added the hook_block instrumentation code, and what index in the tcg arrays to insert it.

### CODE REVIEW

**Comment by rhelmot** [comment]
I don't know how to add a test to your CI setup, but the test is that the following program, assembled with an executable stack, should return 65.

```bash
$ as --32 stuff.s -o stuff.o
$ ld -melf_i386 -z execstack stuff.o -o stuff
$ ./stuff
$ echo $?
65
```

here's stuff.s:

```asm
.intel_syntax noprefix

.global _start
_start:
    mov ebp, esp
    sub ebp, 0x4000
    mov edx, ebp

    lea esi, [self_modifying]
    mov edi, ebp
    mov ecx, 0x2d
    call memcpy
    add ebp, 0x2d
    xor ebx, ebx
    call edx

    mov eax, 1
    int 0x80

memcpy:
    cmp ecx, 0
    je _end
    dec ecx
    mov al, byte ptr [esi+ecx]
    mov byte ptr [edi+ecx], al
    jmp memcpy

_end:
    ret

self_modifying:
    inc ebx
    call $+5
    pop esi
    dec byte ptr [esi+11]
    xor edx, edx
    sub esi, 6
_loop_start:
    cmp edx, 5
    jz _loop_end

    mov edi, ebp
    mov ecx, 0x2d
    lea eax, [memcpy]
    call eax
    inc edx
    add ebp, 0x2d
    mov byte ptr [ebp], 0xc3
    jmp _loop_start

_loop_end:
```

**Comment by rhelmot** [comment]
Added regression test with this testcase.

File: tests/regress/x86_self_modifying.s [hunk 1]
```diff
@@ -0,0 +1,51 @@
+.intel_syntax noprefix
```
Comment (aquynh) [inline tests/regress/x86_self_modifying.s]: a minor thing: would you please add some instructions on how to compile it into this file?

**Comment by aquynh** [comment]
merged, but please send another pull req to add instructions on how to compile the asm file. thanks!

**Pull request merged**